ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Params.hs

```diff
@@ -98,7 +98,9 @@ genAsc = ascFromBits <$> QC.choose (1 :: Word8, maxBound - 1)
 
 genKSD :: QC.Gen (Kcp, Scg, Delta)
 genKSD = sized1 $ \sz -> do
-    d <- QC.choose (0, div sz 4)
+    -- k > 1 so we can ensure an alternative schema loses the density comparison
```

Comment:
I think we should add this as a `PRECONDITION` comment of `uniformAdversarialChain` and also add it as a check to `checkAdversarialRecipe`.

Reply:
Added the precondition to `uniformAdversarialChain`.

I left `checkAdversarialRecipe` unmodified as I don't understand what the precondition on the parameter has to do with the purpose of the function. It seems reasonable to me to want to check if the adversarial schema checks fail when called with smaller k's.

Reply:
> I left `checkAdversarialRecipe` unmodified as I don't understand what the precondition on the parameter has to do with the purpose of the function.

The Haddock on `checkAdversarialRecipe` is `Reject a bad 'AdversarialRecipe'`. If k <= 1, then `uniformAdversarialChain` will not necessarily do what we want (its precondition is violated). So that qualifies as a "bad recipe" and should be rejected.

Agreed?

> It seems reasonable to me to want to check if the adversarial schema checks fail when called with smaller k's.

You can do that by intentionally circumventing `checkAdversarialRecipe` to create an unsafe/unsound `CheckedAdversarialRecipe` and calling `uniformAdversarialChain` with that, yeah?

Reply:
I pushed up tiny commit which I hope you can squash into one of yours. I think it resolves the `checkAdversarialRecipe` part of my suggestion.